src/validators/union.rs

```diff
@@ -134,16 +137,42 @@ impl UnionValidator {
                     _ => {
                         // success should always have an exactness
                         debug_assert_ne!(state.exactness, None);
+
                         let new_exactness = state.exactness.unwrap_or(Exactness::Lax);
-                        // if the new result has higher exactness than the current success, replace it
-                        if success
-                            .as_ref()
-                            .map_or(true, |(_, current_exactness)| *current_exactness < new_exactness)
-                        {
-                            // TODO: is there a possible optimization here, where once there has
-                            // been one success, we turn on strict mode, to avoid unnecessary
-                            // coercions for further validation?
-                            success = Some((new_success, new_exactness));
+
+                        // in the case where the exactness is Strict or Lax, we also check
+                        // the number of fields in the model, and prefer the one with more fields
+                        // which is useful for choosing a more specific model in a union, like a subclass
+                        let new_num_fields: Option<usize> = match choice {
+                            CombinedValidator::Model(x) => {
+                                // for models, we attempt to fetch the number of fields set, as this is a better
+                                // measure of the fit of the input data than the total number of fields on a model
+                                // but we fallback to the total number of fields
+                                new_success
+                                    .getattr(py, intern!(py, "__pydantic_fields_set__"))
```

Comment:
It looks like `__pydantic_fields_set__` is set in `validators/model.rs`. Possibly room for code re-use from there.

Reply:
I think for now, just pulling said data out here makes sense, but one thing we could consider is attaching that data to `state` or something like that so we don't have to pull it off of the result?

Reply:
Then we can decouple the `num_fields` thing from the `Validator` pattern a bit...

Reply:
Thinking about this more - it seems that `state` is more so something that we set pre-validation and then change if needed, so I'm not sure if it's the right place for this data. For now, I'll skip changing `state`, but we can certainly refactor down the line if desired.

I think that might make more sense at a point where we add tiebreaker metrics to other validators.